examples/petsc/areadmplex.c

```diff
@@ -0,0 +1,368 @@
+                            sizeof(ceedresource), NULL); CHKERRQ(ierr);
+  ierr = PetscOptionsBool("-test",
+                          "Testing mode (do not print unless error is large)",
+                          NULL, test_mode, &test_mode, NULL); CHKERRQ(ierr);
+  ierr = PetscOptionsString("-mesh", "Read mesh from file", NULL,
+                            filename, filename, sizeof(filename), &read_mesh);
+  CHKERRQ(ierr);
+  ierr = PetscOptionsEnd(); CHKERRQ(ierr);
+
+  // Setup DM
+  PetscScalar l = 1.0/PetscSqrtReal(3.0); // half edge of the cube
+  if (read_mesh) {
+    ierr = DMPlexCreateFromFile(PETSC_COMM_WORLD, filename, PETSC_TRUE, &dm);
+    CHKERRQ(ierr);
+  } else {
+    // Create the mesh as a 0-refined sphere. This will create a cubic surface, not a box.
+    PetscBool simplex = PETSC_FALSE;
+    ierr = DMPlexCreateSphereMesh(PETSC_COMM_WORLD, topodim, simplex, &dm); CHKERRQ(ierr);
+    CHKERRQ(ierr);
+    ierr = PetscObjectSetName((PetscObject) dm, "Cube"); CHKERRQ(ierr);
+    // Distribute mesh over processes
+    {
+      DM dmDist = NULL;
+      PetscPartitioner part;
+
+      ierr = DMPlexGetPartitioner(dm, &part); CHKERRQ(ierr);
+      ierr = PetscPartitionerSetFromOptions(part); CHKERRQ(ierr);
+      ierr = DMPlexDistribute(dm, 0, NULL, &dmDist); CHKERRQ(ierr);
+      if (dmDist) {
+        ierr = DMDestroy(&dm); CHKERRQ(ierr);
+        dm  = dmDist;
+      }
+    }
+    // View DMPlex via runtime option
+    ierr = DMViewFromOptions(dm, NULL, "-dm_view"); CHKERRQ(ierr);
+  }
+
+  // Create FE
+  ierr = PetscFECreateDefault(PETSC_COMM_SELF, topodim, ncompu, PETSC_FALSE, NULL,
+                              PETSC_DETERMINE, &fe);
+  CHKERRQ(ierr);
+  ierr = DMSetFromOptions(dm); CHKERRQ(ierr);
+  ierr = DMAddField(dm, NULL, (PetscObject)fe); CHKERRQ(ierr);
+  ierr = DMCreateDS(dm); CHKERRQ(ierr);
+  ierr = DMPlexSetClosurePermutationTensor(dm, PETSC_DETERMINE, NULL);
+  CHKERRQ(ierr);
+
+  // Get basis space degree
+  ierr = PetscFEGetBasisSpace(fe, &sp); CHKERRQ(ierr);
+  ierr = PetscSpaceGetDegree(sp, &degree, NULL); CHKERRQ(ierr);
+  ierr = PetscFEDestroy(&fe); CHKERRQ(ierr);
+  if (degree < 1) SETERRQ1(PETSC_COMM_WORLD, PETSC_ERR_ARG_OUTOFRANGE,
+                           "-petscspace_degree %D must be at least 1", degree);
+
+  // Create vectors
+  ierr = DMCreateGlobalVector(dm, &X); CHKERRQ(ierr);
+  ierr = VecGetLocalSize(X, &lsize); CHKERRQ(ierr);
+  ierr = VecGetSize(X, &gsize); CHKERRQ(ierr);
+  ierr = DMCreateLocalVector(dm, &Xloc); CHKERRQ(ierr);
+  ierr = VecGetSize(Xloc, &xlsize); CHKERRQ(ierr);
+  ierr = VecDuplicate(X, &V); CHKERRQ(ierr);
+  ierr = VecDuplicate(Xloc, &Vloc); CHKERRQ(ierr);
+
+  // Set up libCEED
+  CeedInit(ceedresource, &ceed);
+
+  // Print summary
+  P = degree + 1;
+  Q = P + qextra;
+  const char *usedresource;
+  CeedGetResource(ceed, &usedresource);
+  if (!test_mode) {
+    ierr = PetscPrintf(comm,
+                       "\n-- CEED -- libCEED + PETSc Surface Area problem --\n"
```

Comment:
Maybe -- libCEED + PETSc -- Surface Area Problem --
? The 'CEED' out front is strange to me

Reply:
Sure